File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& arr, int k) {
        int n = arr.size();

        map<int,int> mp;
        int sum = 0;
        mp[0] = 1;
        int cnt = 0;
        for(int i=0;i<n;i++)
        {
            sum += arr[i];
            int rem = sum - k;
            if(mp.find(rem) != mp.end()) cnt += mp[rem];
            mp[sum] += 1;
        }

        return cnt;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int n = nums.length;

        HashMap<Integer,Integer> mp = new HashMap<>();

        int sum = 0;

        mp.put(0,1);

        int cnt = 0;

        for(int i=0;i<n;i++)
        {
            sum += nums[i];

            if(mp.containsKey(sum-k)) cnt += mp.get(sum-k);
            

            mp.put(sum, mp.getOrDefault(sum, 0) + 1);
        }

        return cnt;

        

    }
}